nixos-rebuild-ng: skip broken test for Nix < 2.18

sternenseemann opened the pull request:
The mock test for building with remote, but distinct target and build host checks a bunch of internals (how many shell commands are executed, which ones exactly) that are special cased on WITH_NIX_2_18, so the test breaks when that is false. I've opted to simply skip the test for now since the expected call list is very annoying to keep maintained and updated for multiple versions.

This should fix NixOS configurations with config nix.package.version < 2.18. To keep it that way, I've added nixos-rebuild-ng with some nixVersions to passthru.tests that aren't likely to be removed in the short term.


<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [Nixpkgs 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/doc/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/doc/manual/release-notes/rl-2505.section.md) Nixpkgs Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
- [NixOS 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) NixOS Release notes)
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md), [pkgs/README.md](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md), [maintainers/README.md](https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md) and other contributing documentation in corresponding paths.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

sternenseemann committed:
nixos-rebuild-ng: skip broken test for Nix < 2.18

The mock test for building with remote, but distinct target and build
host checks a bunch of internals (how many shell commands are executed,
which ones exactly) that are special cased on WITH_NIX_2_18, so the test
breaks when that is false. I've opted to simply skip the test for now
since the expected call list is very annoying to keep maintained and
updated for multiple versions.

This should fix NixOS configurations with config nix.package.version < 2.18.
To keep it that way, I've added nixos-rebuild-ng with some nixVersions
to passthru.tests that aren't likely to be removed in the short term.

thiagokokada commented:
Need to fix format.

thiagokokada commented:
Also, can you add Lix too?

sternenseemann commented:
> Need to fix format.

CI passes?

thiagokokada commented:
> > Need to fix format.
> 
> CI passes?

No:
```
> ## run mypy > Success: no issues found in 15 source files(B > ## run ruff > All checks passed! > ## run ruff format > Would reformat: /nix/store/fcyjrax10ijn2xs7iz26a9d5vw325jzv-src/tests/test_main.py > 1 file would be reformatted, 14 files already formatted
```

This is not the Nix file formatting, but Python.

sternenseemann committed:
nixos-rebuild-ng: add passthru.tests for lix

Note that the code currently assumes that checking >= 2.18 on the
version of an opaque package is enough which works for many things as it
is the fork point of Lix. However, future changes may need to reevaluate
whether it is enough for every check (should more conditional code be
added).
